tooling/debugger/src/context.rs

```diff
@@ -208,48 +305,56 @@ impl<'a, B: BlackBoxFunctionSolver<FieldElement>> DebugContext<'a, B> {
     /// general, the matching between opcode location and source location is 1
     /// to 1, but due to the compiler inlining functions a single opcode
     /// location may expand to multiple source locations.
-    pub(super) fn get_source_call_stack(&self) -> Vec<(OpcodeLocation, Location)> {
+    pub(super) fn get_source_call_stack(&self) -> Vec<(DebugLocation, Location)> {
         self.get_call_stack()
             .iter()
-            .flat_map(|opcode_location| {
-                self.get_source_location_for_opcode_location(opcode_location)
+            .flat_map(|debug_location| {
+                self.get_source_location_for_debug_location(debug_location)
                     .into_iter()
-                    .map(|source_location| (*opcode_location, source_location))
+                    .map(|source_location| (*debug_location, source_location))
             })
             .collect()
     }
 
     /// Returns the absolute address of the opcode at the given location.
     /// Absolute here means accounting for nested Brillig opcodes in BrilligCall
     /// opcodes.
-    pub fn opcode_location_to_address(&self, location: &OpcodeLocation) -> usize {
-        match location {
-            OpcodeLocation::Acir(acir_index) => self.acir_opcode_addresses[*acir_index],
+    pub fn debug_location_to_address(&self, location: &DebugLocation) -> usize {
+        let circuit_addresses = &self.acir_opcode_addresses[location.circuit_id as usize];
+        match &location.opcode_location {
+            OpcodeLocation::Acir(acir_index) => circuit_addresses[*acir_index],
             OpcodeLocation::Brillig { acir_index, brillig_index } => {
-                self.acir_opcode_addresses[*acir_index] + *brillig_index
+                circuit_addresses[*acir_index] + *brillig_index
             }
         }
     }
 
-    pub fn address_to_opcode_location(&self, address: usize) -> Option<OpcodeLocation> {
-        if address >= *self.acir_opcode_addresses.last().unwrap_or(&0) {
+    pub fn address_to_debug_location(&self, address: usize) -> Option<DebugLocation> {
+        if address >= *self.acir_opcode_addresses.last()?.first()? {
             return None;
         }
-        let location = match self.acir_opcode_addresses.binary_search(&address) {
+        let circuit_id = match self
+            .acir_opcode_addresses
+            .binary_search_by(|addresses| addresses[0].cmp(&address))
+        {
+            Ok(found_index) => found_index,
+            Err(insert_index) => insert_index - 1,
```

Comment:
I'm surprised to see us use the insert index from an error. Don't we build all acir opcode addresses when we create the debugger context? Should this not be an internal debugger error?

Reply:
I see why this is confusing and definitely supports your suggestion about extracting the address mapping to a different type. We're using the insert index because we're doing a binary search of the addresses of the first opcode of each circuit only. This is to find which circuit the address belongs to.

And in the second binary search below this, it's also necessary because we're tracking the addresses of the ACIR opcodes only. Since the program is flattened for the purposes of computing the addresses, Brillig calls effectively "inline" all the Brillig opcodes in the same address space, so there may be holes in the nested vectors of addresses as well.

Reply:
Ok thank you for elaborating. Yeah I do think this would be much easier to interpret if it was extracting into a type.